Keycloak: script for adding new users based on list of emails

frewmack opened the pull request:
Closes #245 

EDIT:
`example.txt`:

```txt
user-alpha@domain.ca
user-bravo@domain.ca

```

Usage:

```
cat example.txt | admin-scripts/add-users-by-email.sh
```

frewmack committed:
Add add-user script to keycloak

frewmack committed:
Fix new password, account for existing users

frewmack committed:
Move script after rebase

frewmack committed:
Add example usage

frewmack committed:
cleanup after rebase

frewmack commented:
Marking this branch as draft. Realized that user IDs need to be pulled from Auth0 for G4RD users, and this isn't implemented here yet.

frewmack committed:
Update script to pull user info from Auth0 API

frewmack committed:
Remove comments

frewmack committed:
Add password option

frewmack committed:
Update docker-compose

frewmack committed:
Update env vars for keycloak staging
- Fix handling of different grant types
- Update .env.sample

kevinlul commented on .env.sample:
```diff
@@ -45,3 +45,4 @@ G4RD_CLIENT_ID=exampleclientid
 G4RD_GRANT_TYPE=granttype
 G4RD_AUTH_METHOD=bearer
 G4RD_TOKEN_URL=https://some-provider.example.com
+G4RD_AUTH0_BASE_URL=https://some-subdomain.auth0.com
```
Trailing newline

kevinlul commented on keycloak/bin/add-users-by-email.sh:
```diff
@@ -0,0 +1,71 @@
+fi
+
+accesstoken=$(
+    echo "${tokenresponse}" | grep -zoP '"access_token":\s*"\K[^\s,]*(?="\s*,)'
```
Are you trying to parse JSON with grep?

frewmack replied:
Yep. The regex matches the value inside the double quotes whose key is "access_token". Using something like [jq](https://stedolan.github.io/jq/) would probably be better, but it's not included in the keycloak container by default.

kevinlul replied:
This script bin is only mounted in development. Is this a script to be used in development, or run from a development machine? If so, we can write this script to assume `jq` exists (heck, even Node.js) and get more reliable parsing there, and use `docker-compose exec keycloak` for the calls to kcadm.

frewmack replied:
My understanding was that this script would be used to semi-automate the process of adding G4RD users to OSMP Keycloak in production. I suppose it does make more sense then to fix the script to run outside of Docker by some trusted admin user in production. I'll rework the script.

frewmack committed:
Fix trailing newline

frewmack committed:
Update script
- Move out of keycloak bin
- Update to use jq to parse JSON
- Script now runs outside of container as admin tool
- Added line to read in .env vars, accounting for nested vars

kevinlul commented on admin-scripts/add-users-by-email.sh:
```diff
@@ -0,0 +1,66 @@
+)
+
+# Get users by email
+file=${1:--}
```
```suggestion
file=${1:-/dev/stdin}
```

kevinlul commented on admin-scripts/add-users-by-email.sh:
```diff
@@ -0,0 +1,66 @@
+        printf "Failed to add user\n"
+    fi
+    set -e
+done < <(cat -- "$file")
```
```suggestion
done < "$file"
```
Avoids starting a subshell just to read a file or standard input

kevinlul commented on admin-scripts/add-users-by-email.sh:
```diff
@@ -0,0 +1,66 @@
+
+    # try to add user to keycloak with this id
+    printf "Adding user with email '%s' and id '%s'... " "$line" "$userid"
+    set +e
+    docker-compose exec keycloak /opt/jboss/keycloak/bin/kcadm.sh create users -s username="$userid" -s email="$line" -s enabled=true -r "${KEYCLOAK_REALM}"
+    if [[ $? -eq 0 ]]; then
```
I believe you can also just do `if command-here` but this is probably more clear

frewmack committed:
Add suggested changes

frewmack committed:
Fix grant type env var

frewmack committed:
Add explicit 'scope' parameter to token req

frewmack committed:
Fix 'not a TTY' error

jennyziyi-xu commented:
![Screen Shot 2022-07-14 at 4 40 05 PM](https://user-images.githubusercontent.com/56279639/179079179-69dc4ec2-7fb4-4552-8418-355357022dba.png)

Hi Mackenzie, this is my example.txt, I was able to run your script and the user with the first email was added to Keycloak successfully! However, the second email was not processed. If I put the second email on the first line, it will be processed. Do I have the correct format for example.txt? It seems only the email on the first line is processed.

frewmack committed:
Add script for obtaining user id from email

frewmack committed:
Fix docker-compose breaking loop; error checking

frewmack committed:
Add federated identity linking to script

jennyziyi-xu commented:
I think `G4RD_CLIENT_SECRET` is missing from `.env.sample`.

frewmack committed:
Update .env.sample

jennyziyi-xu reviewed (approved):
LGTM!!!
